[FLINK-16871][runtime] Make more build time information available at runtime

nielsbasjes opened the pull request:
# [FLINK-16871](https://issues.apache.org/jira/browse/FLINK-16871): Make more build time information available at runtime

## What is the purpose of the change

As discussed [here](https://github.com/apache/flink/pull/11245) additional build time information is needed at runtime. This pull request is to make this specific change a separate thing.

## Brief change log

  - Generate a Version.java file from maven with several properties.
  - Update the git-commit-id-plugin to the latest version (only the mentioned 2.1.14 was problematic). And updated the url in the comment to the git-commit-id-plugin github repo.
  - By fixating the datetime format in the git-commit-id-plugin these can be reliably exposed as returning an `Instant`. This makes it possible to reliably format these timestamps in any form we like. The code has been structured in such a way that as little as possible memory remains allocated. For the latest commit and build time a String version of those timestamps is also available and those are formatted using the Europe/Berlin timezone (the origin of Flink) and thus independent of the locale/timezone of the system that created the build.
  - The .version.properties is no longer generated as this was part of the old solution and it is not used anywhere else in the code base.


## Verifying this change

This change added a few basic test steps.

Verifying the actual values is not possible because after commit and on each build the values will (and should) be different. Also checking if the commit time is not older than X would mean that a working version of the software would fail the build a while later.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **yes: build time plugin is updated**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **no**
  - The serializers: **no**
  - The runtime per-record code paths (performance sensitive): **no**
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: **no**
  - The S3 file system connector: **no**

## Documentation

  - Does this pull request introduce a new feature? **no, not really** A few more getters are now available in the EnvironmentInformation which only affects developers who need this information.
  - If yes, how is the feature documented? **JavaDocs**

nielsbasjes commented:
On my machine the `mvn clean verify` passed.

nielsbasjes commented:
@azagrebin / @wangyang0918 Would you be so kind to have a look at this. 
This is purely the "Environment variable" part we discussed.

azagrebin commented:
Thanks for splitting this @nielsbasjes 

As I understand, the current PR still requires running code generation otherwise the runtime module is broken. I agree that e.g. IntelliJ Idea has a button in maven panel to generate sources which takes some time, at least initially. Nonetheless, this will take time for people to adopt to this.

As mentioned in the [original discussion](https://github.com/apache/flink/pull/11245#discussion_r400035133), we could try approach where we have a `Version.java` file in our code base with "unknown" versions and pack the real versions into the built artefacts.

@zentol might also have further opinion about this.

zentol commented on pom.xml:
```diff
@@ -1811,14 +1811,18 @@ under the License.
 					<artifactId>git-commit-id-plugin</artifactId>
-					<!-- Don't use 2.1.14 as it is incompatible with various maven versions -->
-					<version>2.1.10</version>
+					<version>4.0.0</version>
```
why are we changing the version?

nielsbasjes replied:
I generally update dependencies when the result still works.
In this case there were also a few CVEs that were fixed and more compatibility in building in the various clouds: https://github.com/git-commit-id/git-commit-id-maven-plugin/issues?q=milestone%3A4.0.0

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/EnvironmentInformation.java:
```diff
@@ -41,14 +49,77 @@
+	 * @return Does this built have modifications that have not yet been committed.
+	 */
+	public static boolean getGitDirty() {
+		return GIT_DIRTY;
```
this seems quite Flink developer oriented and I'm not convinced we need it. How would this be exposed to the developer?

nielsbasjes replied:
I agree. Removing it.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/EnvironmentInformation.java:
```diff
@@ -41,14 +49,77 @@
+	 *
+	 * @return The "name:tag" of the Flink docker image.
+	 */
+	public static String getFlinkDockerImageNameAndTag() {
```
I would off on this one; it is currently not clear how this will be used and it will make more sense in the follow-up PR.

nielsbasjes replied:
I agree. Removing it.

zentol commented on flink-runtime/pom.xml:
```diff
@@ -402,6 +402,38 @@ under the License.
+						<replacement><token>@scala.version@</token><value>${scala.binary.version}</value></replacement>
+
+						<replacement><token>@git.build.time@</token><value>${git.build.time}</value></replacement>
+						<replacement><token>@git.build.version@</token><value>${git.build.version}</value></replacement>
```
this isn't used anywhere?

nielsbasjes replied:
I'll remove the ones that are not used.

nielsbasjes commented:
@azagrebin @zentol About generating the Version.java and switching branches.
If you look at the rest of the Flink code base you'll find that in several places things are generated in the `generate-sources` phase. I see a lot of places using things like Protobuf, Avro and an SQL parser that all rely on this same mechanism and the same way of working.

So in general the problems you are addressing already exist in a large scale in the project. You cannot simply switch branch and then successfully run the project without doing a `mvn clean generate-sources`. I see this in almost all of my own projects and also in almost all Apache projects: code needs to be generated before the project can be compiled and tested.

Putting a question back: What implementation approach do you recommend?

zentol commented:
>You cannot simply switch branch and then successfully run the project without doing a mvn clean generate-sources.

This isn't quite accurate. By virtue of us rarely adding code generation and generally not changing it, it _over time_ becomes feasible to switch between the currently supported version branches.
Naturally this goes out the window the moment we add a new code generation, hence this discussion.

> Putting a question back: What implementation approach do you recommend?

So, I only briefly looked into it.

My original idea was to have a normal `Version.java` in the source, to be modified during the build only, but you'd have to jump through quite a few hoops to make this work. (Exclude the vanilla version from the source, add the modified file into a custom directory so IntelliJ doesn't complain, add this directory to the list of source directories). So this isn't really feasible, which is unfortunate, and a bit surprising since the use-case is fairly simple.

One approach we could take though is similar to the old one we had.
Since we're just interested in exposing maven properties we should be able to add a text file into `src/main/resources` containing lines like `PROJECT_VERSION: ${project.version}` and configure maven to use filtering for this resource.
```
<project>
  <build>
    <resources>
        <resource>
            <directory>src/main/resources</directory>
            <filtering>true</filtering>
        </resource>
    </resources>
  </build>
</project>
```

Then just parse it back. This could even be a JSON file that we throw jackson at.

However, I do realize that I may be too strict in regards to code-generation, so I would encourage you to raise the issue on the mailing list in search for supporters.

nielsbasjes commented:
I went back to the generating a .version.properties file which is then loaded as a resource using classloader (which I consider to be brittle and unreliable!). This way you can say 'if no file then use default' as requested. The big downside of this is that someone can (accidentally) change the code by overriding this resource by putting a different one in the classpath (which happens) and IntelliJ still messes things up occasionally. So the runtime code has to catch these problems.

Also in the original code
1. The  .version.properties is generated somewhere into src/main/... instead of target/...
    Why? Generated code (that is not committed) should never be placed in the src/main.
    I changed that to generate to target/
2. A second .version.properties is generated somewhere into flink-dist/src/main/flink-bin/
    Why? The only place the .version.properties is referenced is in the flink-runtime. 
    I consider the overriding these build time variables is a bad idea.
    I removed that.

NOTE: If you run this now from within IntelliJ it still does not generate the files correctly (i.e. IntelliJ does not follow the instructions defined in the pom.xml) which sometimes creates a version file like this (which is handled by the runtime code):
```
project.version=1.11-SNAPSHOT
scala.binary.version=2.11

git.commit.id=${git.commit.id}
git.commit.id.abbrev=${git.commit.id.abbrev}
git.commit.time=${git.commit.time}
git.build.time=${git.build.time}
```

Side note: Looking a the `RevisionInformation` class I found that all uses get that class and then take out the two values (or even only one of them) and throw it away again. Seems too much overhead to me.

nielsbasjes commented:
@zentol The patch I submitted just now does what was originally the case (i.e. the generated properties file) which has been extended with a few additional attributes. Essentially the solution you suggested just now

Although I really think the generated Version.java (as I submitted before) is the best way to do this, I fine with either solution.

Can you please review what is there now (and my previous post, which we both posted at almost exactly the same moment) so this foundation for https://issues.apache.org/jira/browse/FLINK-15794 can be used to begin that change.

Thanks.

nielsbasjes commented:
@zentol As suggested I have posted the question to the dev mailing list: http://mail-archives.apache.org/mod_mbox/flink-dev/202004.mbox/%3CCADoiZqq_VQEn57K3RwNv%3DDXgDWFf%3D5d6b9rX2odJh_PRmj7PsA%40mail.gmail.com%3E

nielsbasjes commented:
@zentol I have been thinking about everything that was said and I think I found a way that is in line with what was said in the mailing list. 
In my opinion the file must be present and have valid values. Because these values will be used to load the correct docker image: If it has invalid values it is unsafe to continue as this will lead to bad testing.

In line with the remark by @tillrohrmann if the file is missing or badly generated the system will throw a
`java.lang.IllegalStateException: The file .flink-runtime.version.properties has not been generated correctly. You MUST run 'mvn generate-sources' in the flink-runtime module.`

So the file is either valid or will lead to an exception.

Please indicate if this is an acceptable solution.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/EnvironmentInformation.java:
```diff
@@ -41,42 +45,148 @@
+
+		private String getProperty(Properties properties, String key) {
+			String value = properties.getProperty(key);
+			if (value == null || value.charAt(0) == '$') {
```
The `$` branch means that it will fail if I just run it in the IDE; instead we should return the unknown value.

nielsbasjes replied:
Ok, I'll update that behaviour.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/EnvironmentInformation.java:
```diff
@@ -41,42 +45,148 @@
+		private static final String  DEFAULT_TIME_STRING  = "1970-01-01T00:00:00Z";
+
+
+		private String  projectVersion    = UNKNOWN;
+		private String  scalaVersion      = UNKNOWN;
+		private Instant gitBuildTime      = DEFAULT_TIME_INSTANT;
+		private String  gitBuildTimeStr   = DEFAULT_TIME_STRING;
+		private String  gitCommitId       = UNKNOWN;
+		private String  gitCommitIdAbbrev = UNKNOWN;
+		private Instant gitCommitTime     = DEFAULT_TIME_INSTANT;
+		private String  gitCommitTimeStr  = DEFAULT_TIME_STRING;
```
```suggestion
		private String  projectVersion = UNKNOWN;
		private String  scalaVersion = UNKNOWN;
		private Instant gitBuildTime = DEFAULT_TIME_INSTANT;
		private String  gitBuildTimeStr = DEFAULT_TIME_STRING;
		private String  gitCommitId = UNKNOWN;
		private String  gitCommitIdAbbrev = UNKNOWN;
		private Instant gitCommitTime = DEFAULT_TIME_INSTANT;
		private String  gitCommitTimeStr = DEFAULT_TIME_STRING;
```

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/EnvironmentInformation.java:
```diff
@@ -41,42 +45,148 @@
+	private static final class Versions {
+		private static final Instant DEFAULT_TIME_INSTANT = Instant.EPOCH;
+		private static final String  DEFAULT_TIME_STRING  = "1970-01-01T00:00:00Z";
+
+
```
```suggestion

```

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/EnvironmentInformation.java:
```diff
@@ -41,42 +45,148 @@
+		}
+
+		public Versions() {
+
```
```suggestion
```

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/EnvironmentInformation.java:
```diff
@@ -41,42 +45,148 @@
+		public Versions() {
+
+			ClassLoader classLoader = EnvironmentInformation.class.getClassLoader();
+			try(InputStream propFile = classLoader.getResourceAsStream(PROP_FILE)) {
```
```suggestion
			try (InputStream propFile = classLoader.getResourceAsStream(PROP_FILE)) {
```

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/EnvironmentInformation.java:
```diff
@@ -41,42 +45,148 @@
+			}
+			catch(IOException ioe) {
+				if (LOG.isDebugEnabled()) {
+					LOG.debug("Cannot determine code revision: Unable to read version property file.: {}", ioe.getMessage());
```
why this distinction?

nielsbasjes replied:
Fixed

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/EnvironmentInformation.java:
```diff
@@ -367,5 +478,13 @@ public RevisionInformation(String commitId, String commitDate) {
 		}
+
+		@Override
+		public String toString() {
```
Why are we adding this?

nielsbasjes replied:
I added this because while debugging IntelliJ uses the toString method to show the content of a record ... which is messy if it falls back to the Object.toString(). I'll remove it.

nielsbasjes commented:
So what it does now:
- NO properties file --> All values are "unknown".
- Poorly generated file (usually from IDE): The project and scala versions are commonly correct. The 'git' properties are still default. This is because the maven git plugin is not run by the IDE so these properties are not set.
- A cleanly generated file (usually by maven): Everything is correct.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/EnvironmentInformation.java:
```diff
@@ -41,42 +45,143 @@
+	}
+
+	private static final class Versions {
+		private static final Instant DEFAULT_TIME_INSTANT = Instant.EPOCH;
+		private static final String  DEFAULT_TIME_STRING = "1970-01-01T00:00:00+0000";
+		private static final String  UNKNOWN_COMMIT_ID = "DecafC0ffeeD0d0F00d";
+		private static final String  UNKNOWN_COMMIT_ID_ABBREV = "DeadD0d0";
+		private String  projectVersion = UNKNOWN;
+		private String  scalaVersion = UNKNOWN;
+		private Instant gitBuildTime = DEFAULT_TIME_INSTANT;
+		private String  gitBuildTimeStr = DEFAULT_TIME_STRING;
+		private String  gitCommitId = UNKNOWN_COMMIT_ID;
+		private String  gitCommitIdAbbrev = UNKNOWN_COMMIT_ID_ABBREV;
+		private Instant gitCommitTime = DEFAULT_TIME_INSTANT;
+		private String  gitCommitTimeStr = DEFAULT_TIME_STRING;
```
```suggestion
		private static final Instant DEFAULT_TIME_INSTANT = Instant.EPOCH;
		private static final String DEFAULT_TIME_STRING = "1970-01-01T00:00:00+0000";
		private static final String UNKNOWN_COMMIT_ID = "DecafC0ffeeD0d0F00d";
		private static final String UNKNOWN_COMMIT_ID_ABBREV = "DeadD0d0";
		private String projectVersion = UNKNOWN;
		private String scalaVersion = UNKNOWN;
		private Instant gitBuildTime = DEFAULT_TIME_INSTANT;
		private String gitBuildTimeStr = DEFAULT_TIME_STRING;
		private String gitCommitId = UNKNOWN_COMMIT_ID;
		private String gitCommitIdAbbrev = UNKNOWN_COMMIT_ID_ABBREV;
		private Instant gitCommitTime = DEFAULT_TIME_INSTANT;
		private String gitCommitTimeStr = DEFAULT_TIME_STRING;
```

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/EnvironmentInformation.java:
```diff
@@ -41,42 +45,143 @@
+		private static final Versions INSTANCE = new Versions();
+	}
+
+	public static Versions getVersionsInstance() {
```
private?

zentol replied:
(since `Versions` is private anyway so you couldn't do anything interesting with it)

nielsbasjes replied:
Ah, yes. I'll fix that.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/EnvironmentInformation.java:
```diff
@@ -41,42 +45,143 @@
-		return new RevisionInformation(revision, commitDate);
+	}
+
+	private static final class VersionsHolder {
```
We usually don't encapsulate singletons. What is the goal behind the holder? (I haven't seen this before)

nielsbasjes replied:
It is a trick to ensure the instantiation of the singleton does not incur any race conditions.
This is a safe solution because the class instantiation of the VersionsHolder is guaranteed to be single threaded. 
Another way to do this is to make the getVersionsInstance synchronized which makes it needlessly slow.

nielsbasjes replied:
This is also the reason why the VersionsHolder is private.

zentol replied:
Are you saying that having `private static final Versions INSTANCE = new Versions();` directly in `EnvironmentInformation` would _not_ be thread-safe?

nielsbasjes replied:
No that is thread safe too. The way I did it the initialization is done on first use instead of on class loading. Which do you prefer?

zentol replied:
Ah, got it. hmm...we usually use the class-loading one I believe, but I suppose  it's fine either way

nielsbasjes commented:
I did a rebase + squash to the latest master branch.
